server/sonar-server/src/main/java/org/sonar/server/activity/index/ActivityIndex.java

```diff
@@ -0,0 +1,100 @@
+import org.elasticsearch.action.search.SearchRequestBuilder;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.index.query.AndQueryBuilder;
+import org.elasticsearch.index.query.OrQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.search.sort.SortOrder;
+import org.sonar.core.util.NonNullInputFunction;
+import org.sonar.server.es.BaseIndex;
+import org.sonar.server.es.EsClient;
+import org.sonar.server.es.SearchOptions;
+import org.sonar.server.es.SearchResult;
+
+public class ActivityIndex extends BaseIndex {
+
+  /**
+   * Convert an Elasticsearch result (a map) to an {@link org.sonar.server.activity.index.ActivityDoc}. It's
+   * used for {@link org.sonar.server.es.SearchResult}.
+   */
+  private static final Function<Map<String, Object>, ActivityDoc> DOC_CONVERTER = new NonNullInputFunction<Map<String, Object>, ActivityDoc>() {
+    @Override
+    protected ActivityDoc doApply(Map<String, Object> input) {
+      return new ActivityDoc(input);
+    }
+  };
+
+  public ActivityIndex(EsClient esClient) {
+    super(esClient);
+  }
+
+  public SearchResult<ActivityDoc> search(ActivityQuery query, SearchOptions options) {
+    SearchResponse response = doSearch(query, options);
+    return new SearchResult<>(response, DOC_CONVERTER);
+  }
+
+  public SearchResponse doSearch(ActivityQuery query, SearchOptions options) {
+    SearchRequestBuilder requestBuilder = getClient()
+      .prepareSearch(ActivityIndexDefinition.INDEX)
+      .setTypes(ActivityIndexDefinition.TYPE);
+
+    requestBuilder.setFrom(options.getOffset());
+    requestBuilder.setSize(options.getLimit());
+    requestBuilder.addSort(ActivityIndexDefinition.FIELD_CREATED_AT, SortOrder.DESC);
+
+    AndQueryBuilder filter = QueryBuilders.andQuery();
+    if (!query.getTypes().isEmpty()) {
+      OrQueryBuilder typeQuery = QueryBuilders.orQuery();
+      for (String type : query.getTypes()) {
+        typeQuery.add(QueryBuilders.termQuery(ActivityIndexDefinition.FIELD_TYPE, type));
+      }
+      filter.add(typeQuery);
+    }
+
+    if (!query.getDataOrFilters().isEmpty()) {
+      for (Map.Entry<String, Object> entry : query.getDataOrFilters().entrySet()) {
+        OrQueryBuilder orQuery = QueryBuilders.orQuery();
+        orQuery.add(QueryBuilders.nestedQuery(ActivityIndexDefinition.FIELD_DETAILS,
+          QueryBuilders.termQuery(ActivityIndexDefinition.FIELD_DETAILS + "." + entry.getKey(), entry.getValue())));
+        filter.add(orQuery);
+      }
+    }
+
+    Date since = query.getSince();
+    if (since != null) {
+      filter.add(QueryBuilders.rangeQuery(ActivityIndexDefinition.FIELD_CREATED_AT)
+        .gt(since));
+    }
+    Date to = query.getTo();
+    if (to != null) {
+      filter.add(QueryBuilders.rangeQuery(ActivityIndexDefinition.FIELD_CREATED_AT)
+        .lt(to));
+    }
+
+    requestBuilder.setQuery(QueryBuilders.filteredQuery(QueryBuilders.matchAllQuery(), filter));
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this use of "filteredQuery"; it is deprecated. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)